File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/npcs.ts
/**
 * Campaign-scoped NPC records: the DM tracks a private agenda alongside a
 * player-visible public status. See shared.ts for the ownership model.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { authenticate, isActor, isApiResult, findCampaign, requireParticipant } from './shared.ts';

type NpcRow = {
  campaign_id: string;
  npc_id: string;
  name: string;
  agenda: string;
  public_status: string;
};

const NPC_NOT_FOUND: ApiResult = { status: 404, body: { error: 'npc not found' } };

function findNpc(db: ReturnType<typeof getDb>, campaignId: string, npcId: string): NpcRow | ApiResult {
  const npc = db
    .prepare('SELECT campaign_id, npc_id, name, agenda, public_status FROM play_campaign_npcs WHERE campaign_id = ? AND npc_id = ?')
    .get(campaignId, npcId) as NpcRow | undefined;
  return npc ?? NPC_NOT_FOUND;
}

export function createCampaignNpc(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may create npcs' } };
  }

  const npcId = body.npc_id;
  if (typeof npcId !== 'string' || npcId.length === 0) {
    return { status: 400, body: { error: 'npc_id must be a non-empty string' } };
  }

  const name = body.name;
  if (typeof name !== 'string' || name.length === 0) {
    return { status: 400, body: { error: 'name must be a non-empty string' } };
  }

  const agenda = body.agenda;
  if (typeof agenda !== 'string' || agenda.length === 0) {
    return { status: 400, body: { error: 'agenda must be a non-empty string' } };
  }

  const publicStatus = body.public_status;
  if (typeof publicStatus !== 'string' || publicStatus.length === 0) {
    return { status: 400, body: { error: 'public_status must be a non-empty string' } };
  }

  const existing = findNpc(db, campaignId, npcId);
  if (!isApiResult(existing)) {
    return { status: 409, body: { error: 'npc_id already exists in this campaign' } };
  }

  db.prepare(
    'INSERT INTO play_campaign_npcs (campaign_id, npc_id, name, agenda, public_status) VALUES (?, ?, ?, ?, ?)',
  ).run(campaignId, npcId, name, agenda, publicStatus);

  return {
    status: 201,
    body: { npc_id: npcId, name, agenda, public_status: publicStatus },
  };
}

export function updateCampaignNpcAgenda(
  authHeader: string | undefined,
  campaignId: string,
  npcId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may update npc agendas' } };
  }

  const agenda = body.agenda;
  if (typeof agenda !== 'string' || agenda.length === 0) {
    return { status: 400, body: { error: 'agenda must be a non-empty string' } };
  }

  const publicStatus = body.public_status;
  if (typeof publicStatus !== 'string' || publicStatus.length === 0) {
    return { status: 400, body: { error: 'public_status must be a non-empty string' } };
  }

  const npc = findNpc(db, campaignId, npcId);
  if (isApiResult(npc)) return npc;

  db.prepare(
    'UPDATE play_campaign_npcs SET agenda = ?, public_status = ? WHERE campaign_id = ? AND npc_id = ?',
  ).run(agenda, publicStatus, campaignId, npcId);

  return {
    status: 200,
    body: { npc_id: npc.npc_id, name: npc.name, agenda, public_status: publicStatus },
  };
}

export function getCampaignNpc(authHeader: string | undefined, campaignId: string, npcId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const npc = findNpc(db, campaignId, npcId);
  if (isApiResult(npc)) return npc;

  if (actor.username === campaign.owner) {
    return {
      status: 200,
      body: { npc_id: npc.npc_id, name: npc.name, agenda: npc.agenda, public_status: npc.public_status },
    };
  }

  return {
    status: 200,
    body: { npc_id: npc.npc_id, name: npc.name, public_status: npc.public_status },
  };
}

type DialogueRow = {
  dialogue_id: string;
  speaker: string;
  text: string;
  visibility: string;
};

function nextDialogueSequence(db: ReturnType<typeof getDb>, campaignId: string, npcId: string): number {
  const row = db
    .prepare(
      'SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_npc_dialogue WHERE campaign_id = ? AND npc_id = ?',
    )
    .get(campaignId, npcId) as { max_sequence: number };
  return row.max_sequence + 1;
}

export function addNpcDialogue(
  authHeader: string | undefined,
  campaignId: string,
  npcId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may append npc dialogue' } };
  }

  const npc = findNpc(db, campaignId, npcId);
  if (isApiResult(npc)) return npc;

  const dialogueId = body.dialogue_id;
  if (typeof dialogueId !== 'string' || dialogueId.length === 0) {
    return { status: 400, body: { error: 'dialogue_id must be a non-empty string' } };
  }

  const speaker = body.speaker;
  if (typeof speaker !== 'string' || speaker.length === 0) {
    return { status: 400, body: { error: 'speaker must be a non-empty string' } };
  }

  const text = body.text;
  if (typeof text !== 'string' || text.length === 0) {
    return { status: 400, body: { error: 'text must be a non-empty string' } };
  }

  const visibility = body.visibility;
  if (visibility !== 'public' && visibility !== 'private') {
    return { status: 400, body: { error: "visibility must be exactly 'public' or 'private'" } };
  }

  const existing = db
    .prepare(
      'SELECT dialogue_id FROM play_campaign_npc_dialogue WHERE campaign_id = ? AND npc_id = ? AND dialogue_id = ?',
    )
    .get(campaignId, npcId, dialogueId);
  if (existing) {
    return { status: 409, body: { error: 'dialogue_id already exists for this npc' } };
  }

  const sequence = nextDialogueSequence(db, campaignId, npcId);
  db.prepare(
    `INSERT INTO play_campaign_npc_dialogue
       (campaign_id, npc_id, sequence, dialogue_id, speaker, text, visibility)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
  ).run(campaignId, npcId, sequence, dialogueId, speaker, text, visibility);

  return {
    status: 201,
    body: { dialogue_id: dialogueId, speaker, text, visibility },
  };
}

export function getNpcDialogue(authHeader: string | undefined, campaignId: string, npcId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const npc = findNpc(db, campaignId, npcId);
  if (isApiResult(npc)) return npc;

  const rows = db
    .prepare(
      `SELECT dialogue_id, speaker, text, visibility
       FROM play_campaign_npc_dialogue
       WHERE campaign_id = ? AND npc_id = ?
       ORDER BY sequence ASC`,
    )
    .all(campaignId, npcId) as DialogueRow[];

  const entries = actor.username === campaign.owner ? rows : rows.filter((row) => row.visibility === 'public');

  return {
    status: 200,
    body: {
      npc_id: npc.npc_id,
      entries: entries.map((entry) => ({
        dialogue_id: entry.dialogue_id,
        speaker: entry.speaker,
        text: entry.text,
        visibility: entry.visibility,
      })),
    },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/npcs.rb
# Campaign-scoped NPC records for the /v1/play surface: the DM tracks a
# private agenda alongside a player-visible public status.

def find_play_npc!(campaign_id, npc_id)
  npc = db.execute(
    'SELECT * FROM play_npcs WHERE campaign_id = ? AND npc_id = ?',
    [campaign_id, npc_id]
  ).first
  halt 404, { error: 'npc not found' }.to_json unless npc
  npc
end

def play_npc_dm_payload(npc)
  {
    npc_id: npc['npc_id'],
    name: npc['name'],
    agenda: npc['agenda'],
    public_status: npc['public_status']
  }
end

def play_npc_player_payload(npc)
  {
    npc_id: npc['npc_id'],
    name: npc['name'],
    public_status: npc['public_status']
  }
end

post '/v1/play/campaigns/:id/npcs' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  npc_id = body['npc_id']
  name = body['name']
  agenda = body['agenda']
  public_status = body['public_status']

  halt 400, { error: 'invalid npc_id' }.to_json unless npc_id.is_a?(String) && !npc_id.empty?
  halt 400, { error: 'invalid name' }.to_json unless name.is_a?(String) && !name.empty?
  halt 400, { error: 'invalid agenda' }.to_json unless agenda.is_a?(String) && !agenda.empty?
  halt 400, { error: 'invalid public_status' }.to_json unless public_status.is_a?(String) && !public_status.empty?

  existing = db.execute(
    'SELECT 1 FROM play_npcs WHERE campaign_id = ? AND npc_id = ?',
    [campaign['id'], npc_id]
  ).first
  halt 409, { error: 'npc_id already exists' }.to_json if existing

  db.execute(
    'INSERT INTO play_npcs (campaign_id, npc_id, name, agenda, public_status) VALUES (?, ?, ?, ?, ?)',
    [campaign['id'], npc_id, name, agenda, public_status]
  )

  status 201
  { npc_id: npc_id, name: name, agenda: agenda, public_status: public_status }.to_json
end

put '/v1/play/campaigns/:id/npcs/:npc_id/agenda' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  npc = find_play_npc!(campaign['id'], params[:npc_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  agenda = body['agenda']
  public_status = body['public_status']

  halt 400, { error: 'invalid agenda' }.to_json unless agenda.is_a?(String) && !agenda.empty?
  halt 400, { error: 'invalid public_status' }.to_json unless public_status.is_a?(String) && !public_status.empty?

  db.execute(
    'UPDATE play_npcs SET agenda = ?, public_status = ? WHERE campaign_id = ? AND npc_id = ?',
    [agenda, public_status, campaign['id'], npc['npc_id']]
  )

  { npc_id: npc['npc_id'], name: npc['name'], agenda: agenda, public_status: public_status }.to_json
end

get '/v1/play/campaigns/:id/npcs/:npc_id' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  npc = find_play_npc!(campaign['id'], params[:npc_id])

  (is_owner ? play_npc_dm_payload(npc) : play_npc_player_payload(npc)).to_json
end

post '/v1/play/campaigns/:id/npcs/:npc_id/dialogue' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  npc = find_play_npc!(campaign['id'], params[:npc_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  dialogue_id = body['dialogue_id']
  speaker = body['speaker']
  text = body['text']
  visibility = body['visibility']

  halt 400, { error: 'invalid dialogue_id' }.to_json unless dialogue_id.is_a?(String) && !dialogue_id.empty?
  halt 400, { error: 'invalid speaker' }.to_json unless speaker.is_a?(String) && !speaker.empty?
  halt 400, { error: 'invalid text' }.to_json unless text.is_a?(String) && !text.empty?
  halt 400, { error: 'invalid visibility' }.to_json unless %w[public private].include?(visibility)

  existing = db.execute(
    'SELECT 1 FROM play_npc_dialogue WHERE campaign_id = ? AND npc_id = ? AND dialogue_id = ?',
    [campaign['id'], npc['npc_id'], dialogue_id]
  ).first
  halt 409, { error: 'dialogue_id already exists' }.to_json if existing

  next_sequence = db.execute(
    'SELECT COALESCE(MAX(sequence), -1) + 1 AS seq FROM play_npc_dialogue WHERE campaign_id = ? AND npc_id = ?',
    [campaign['id'], npc['npc_id']]
  ).first['seq']

  db.execute(
    'INSERT INTO play_npc_dialogue (campaign_id, npc_id, sequence, dialogue_id, speaker, text, visibility) VALUES (?, ?, ?, ?, ?, ?, ?)',
    [campaign['id'], npc['npc_id'], next_sequence, dialogue_id, speaker, text, visibility]
  )

  status 201
  { dialogue_id: dialogue_id, speaker: speaker, text: text, visibility: visibility }.to_json
end

get '/v1/play/campaigns/:id/npcs/:npc_id/dialogue' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  npc = find_play_npc!(campaign['id'], params[:npc_id])

  rows = db.execute(
    'SELECT * FROM play_npc_dialogue WHERE campaign_id = ? AND npc_id = ? ORDER BY sequence ASC',
    [campaign['id'], npc['npc_id']]
  )
  rows = rows.select { |row| row['visibility'] == 'public' } unless is_owner

  entries = rows.map do |row|
    {
      dialogue_id: row['dialogue_id'],
      speaker: row['speaker'],
      text: row['text'],
      visibility: row['visibility']
    }
  end

  { npc_id: npc['npc_id'], entries: entries }.to_json
end
